### DISCUSSION TITLE
Save and load Canvas via JS

### DISCUSSION THREAD

**Discussion by jedimcclain** [excalidraw/excalidraw#7589] [Q&A]
I have excalidraw loaded up in a basic html page. I am using the Browser integration since I am not very familiar with react and the rest of my app is just html and JS.  All built in functionally works as expected.

I am trying to get the JSON of the canvas using JavaScript. I read that the canvas is stored in local storage under the key 'excalidraw '. I see on the production page this is the case, however, when I run on my page there is nothing in local storage. 

Is there a different way I can get the JSON for the canvas? Or am I missing something to turn on local storage? 

My next question is how do I initialize  excalidraw once  if I have the JSON data?

Thanks in advance for any help or pointers,
Scott

For completeness here is my html page:

<!DOCTYPE html>
<html>
<head>
    <title>Excalidraw</title>
    <meta charset="UTF-8" />
    <link href="~/js/styles.css" rel="stylesheet" />
    <script src="https://unpkg.com/react@18.2.0/umd/react.development.js"></script>
    <script src="https://unpkg.com/react-dom@18.2.0/umd/react-dom.development.js"></script>

    <link href="~/lib/excalidraw/17.1/dev/index.css" rel="stylesheet" />

</head>

<body>
    <div class="row">
        <button type="button" onclick="onLoad()" class="btn btn-secondary">Load</button>
        <button type="button" onclick="onSave()" class="btn btn-primary">Save</button>
    </div>
    <div class="container">
        <div id="app"></div>
    </div>
    <script type="module">
        import * as ExcalidrawLib from "/lib/excalidraw/17.1/dev/index.js";
            console.log("EXCALIDRAW LIB", ExcalidrawLib);
        window.ExcalidrawLib = ExcalidrawLib;
    </script>

    
    <script defer type="module" src="./js/initExcalidraw.js"></script>

    <script>
        function onSave() {
            //This is empty
            const serializedData = localStorage.getItem('excalidraw');
        }

        function onLoad(){
           
        }

    </script>

</body>
</html>

#my initial Data:------------------------------------------------------------------------------------
export default {
    elements: [
        {
            type: "rectangle",
            version: 141,
            versionNonce: 361174001,
            isDeleted: false,
            id: "oDVXy8D6rom3H1-LLH2-f",
            fillStyle: "hachure",
            strokeWidth: 1,
            strokeStyle: "solid",
            roughness: 1,
            opacity: 100,
            angle: 0,
            x: 100.50390625,
            y: 93.67578125,
            strokeColor: "#000000",
            backgroundColor: "transparent",
            width: 186.47265625,
            height: 141.9765625,
            seed: 1968410350,
            groupIds: []
        },
        {
            id: "-xMIs_0jIFqvpx-R9UnaG",
            type: "ellipse",
            x: 300.5703125,
            y: 190.69140625,
            width: 198.21875,
            height: 129.51171875,
            angle: 0,
            strokeColor: "#000000",
            backgroundColor: "transparent",
            fillStyle: "hachure",
            strokeWidth: 1,
            strokeStyle: "solid",
            roughness: 1,
            opacity: 100,
            groupIds: [],
            seed: 957947807,
            version: 47,
            versionNonce: 1128618623,
            isDeleted: false
        }
    ],
    appState: { viewBackgroundColor: "#AFEEEE", currentItemFontFamily: 1 }
};

#and my initExcalidraw.js---------------------------------------------------------
/*eslint-disable */

const App = () => {
    const [excalidrawAPI, setExcalidrawAPI] = React.useState(null);
    const excalidrawWrapperRef = React.useRef(null);
    const [dimensions, setDimensions] = React.useState({
        width: undefined,
        height: undefined,
    });

    const [viewModeEnabled, setViewModeEnabled] = React.useState(false);
    const [zenModeEnabled, setZenModeEnabled] = React.useState(true);
    const [gridModeEnabled, setGridModeEnabled] = React.useState(false);

    React.useEffect(() => {
        setDimensions({
            width: excalidrawWrapperRef.current.getBoundingClientRect().width,
            height: excalidrawWrapperRef.current.getBoundingClientRect().height,
        });
        const onResize = () => {
            setDimensions({
                width: excalidrawWrapperRef.current.getBoundingClientRect().width,
                height: excalidrawWrapperRef.current.getBoundingClientRect().height,
            });
        };

        window.addEventListener("resize", onResize);

        return () => window.removeEventListener("resize", onResize);
    }, [excalidrawWrapperRef]);

    return React.createElement(
        React.Fragment,
        null,
        React.createElement(
            "div",
            {
                className: "excalidraw-wrapper",
                ref: excalidrawWrapperRef,
            },
            React.createElement(ExcalidrawLib.Excalidraw, {
                excalidrawAPI: (api) => setExcalidrawAPI(api),
                zenModeEnabled,
                gridModeEnabled,
                viewModeEnabled
            },
                React.createElement(ExcalidrawLib.MainMenu, {})
            )
        )
    );
};

const excalidrawWrapper = document.getElementById("app");
const root = ReactDOM.createRoot(excalidrawWrapper);
root.render(React.createElement(App));

**Comment by ad1992** [comment]
Hi @jedimcclain you can follow these [steps](https://docs.excalidraw.com/docs/@excalidraw/excalidraw/integration#browser) to integrate in browser, and you can use version `0.17.1`

**Reply by jedimcclain** [reply]
My issues, I guess is the example you referend to has never worked for me . I of course started with that. I get an error:
Uncaught TypeError: Cannot read properties of undefined (reading 'jsxs')
    at eval (icons.tsx:74:41)
    at ../../components/icons.tsx (excalidraw.development.js:3588:1)
    at __webpack_require__ (excalidraw.development.js:5614:42)

This is after copy/pasting the example browser integration without any changes in to a basic hosted html page with out any other references. It has the references as listed in the example and there are not errors in finding these packages.

As a result after some work I got the code that I posted previously to work. I just could not access the excalidraw object to call any methods or get any properties or see the JSON data.

**Reply by ad1992** [reply]
@jedimcclain which version are you using? Have you tried with`0.17.1` ?

**Reply by jedimcclain** [reply]
Hi @ad1992 ,

I can get 17.1 to work if I integrate like in my first post. The issue is the JSON data was not showing up in local data. I am not sure where it was storing it. 

If I use the example exactly like on the integration page and it references: https://unpkg.com/@excalidraw/excalidraw/dist/excalidraw.development.js  which I assume is a later version. 

I am not picky on what version I use. I just want to load up in a generic HTML page and save/load a canvas via JavaScript at this point.

Thanks in advance to any pointers.
